notebooks/02_model/vowpal_wabbit_deep_dive.ipynb

```diff
@@ -0,0 +1,737 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<i>Copyright (c) Microsoft Corporation. All rights reserved.</i>\n",
+    "\n",
+    "<i>Licensed under the MIT License.</i>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<center><h1>Using Vowpal Wabbit for Recommendations</h1>\n",
+    "\n",
+    "<img src=\"https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/logo_assets/vowpal-wabbits-github-logo.png?raw=true\" height=\"30%\" width=\"30%\" alt=\"Vowpal Wabbit\">\n",
+    "</center>\n",
+    "\n",
+    "[Vowpal Wabbit](https://github.com/VowpalWabbit/vowpal_wabbit) is a fast online machine learning library that implements several algorithms relevant to the recommendation use case.\n",
+    "\n",
+    "The main advantage of Vowpal Wabbit (VW) is that training is done in an online fashion typically using Stochastic Gradient Descent or similar variants, which allows it to scale well to very large datasets. Additionally, it is optimized to run very quickly and can support distributed training scenarios for extremely large datasets.\n",
+    "\n",
+    "In this notebook we demonstrate how to use the VW library to generate recommendations on the [Movielens](https://grouplens.org/datasets/movielens/) dataset.\n",
+    "\n",
+    "Several things are worth noting in how VW is being used in this notebook. By leveraging an Azure Data Science Virtual Machine ([DSVM](https://azure.microsoft.com/en-us/services/virtual-machines/data-science-virtual-machines/)), VW comes pre-installed and can be used directly from the command line. There are also python bindings to be able to use VW within a python environment and even a wrapper conforming to the SciKit-Learn Estimator API. However, the python bindings must be installed as an additional python package with Boost dependencies, so for simplicity's sake execution of VW is done via a subprocess call mimicking what would happen from the command line execution of the model.\n",
```

Comment:
Right now, we are not installing VW, we are assuming that a person is using a DSVM. How difficult is to add an installation option with the code generator that we have? https://github.com/Microsoft/Recommenders/blob/staging/scripts/generate_conda_file.sh

Reply:
installation is a bit tricky due to a dependency on boost libraries. these can be installed via conda, however it is quite challenging to ensure the correct version is leveraged during building vw (currently does not work with python3). this is an ongoing issue with vw and is something I'm working with others on in that repo. Once, it's resolved we can add that to the conda yaml generator.

Reply:
can you elaborate a little bit on this? what is the exact bottleneck with boost? when this will be solved? are you directly working on VW repo?

Reply:
sure here are the problems:
- current pip install vowpalwabbit works with system installations of boost on python 2 and 3 
   - pip version is slightly outdated
   - requires system level installation, and potential renaming of libraries
- current repo code changed the python build approach and only works with system installations of boost on python 2 (not 3)
  - i'm gathering notes to explain how to replicate this and post an issue
- neither approach works well with conda, but ideally the boost dependencies could be loaded into a conda env, and vw could be build from conda-forge or through pip under conda (now that conda plays a bit nicer with pip)
 
so my hope is that we can work through these issues in the VW repo (I've been contributing to the python bindings there) and resolve the issues there. but as it stands there's no clear timeline